File: tests/smoke-test.py
#!/usr/bin/env python3
"""
LED Wall Calculator — Smoke Test
Runs basic validation checks after code changes.
Usage: python3 tests/smoke-test.py
"""

import json
import os
import re
import sys
from html.parser import HTMLParser

# Paths relative to project root
PROJECT_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
INDEX_HTML = os.path.join(PROJECT_ROOT, "index.html")
VERSION_JSON = os.path.join(PROJECT_ROOT, "version.json")

# External JS module directories to scan
JS_MODULE_DIRS = ["core", "specs", "layouts", "structure", "interact", "nav", "export", "config", "screens"]


def get_all_js_files():
    """Return list of all extracted .js module files."""
    js_files = []
    for d in JS_MODULE_DIRS:
        dir_path = os.path.join(PROJECT_ROOT, d)
        if os.path.isdir(dir_path):
            for fname in sorted(os.listdir(dir_path)):
                if fname.endswith(".js"):
                    js_files.append(os.path.join(dir_path, fname))
    return js_files


def read_all_js_content():
    """Read JS from index.html inline scripts + all extracted .js files."""
    content = ""
    if os.path.exists(INDEX_HTML):
        with open(INDEX_HTML, "r") as f:
            content = f.read()
    for js_file in get_all_js_files():
        with open(js_file, "r") as f:
            content += "\n" + f.read()
    return content

passed = 0
failed = 0
warnings = 0


def ok(msg):
    global passed
    passed += 1
    print(f"  PASS  {msg}")


def fail(msg):
    global failed
    failed += 1
    print(f"  FAIL  {msg}")


def warn(msg):
    global warnings
    warnings += 1
    print(f"  WARN  {msg}")


# ---------------------------------------------------------------------------
# 1. version.json validation
# ---------------------------------------------------------------------------
def check_version_json():
    print("\n--- version.json ---")

    if not os.path.exists(VERSION_JSON):
        fail("version.json not found")
        return None

    try:
        with open(VERSION_JSON, "r") as f:
            data = json.load(f)
    except json.JSONDecodeError as e:
        fail(f"version.json is not valid JSON: {e}")
        return None

    ok("version.json is valid JSON")

    version = data.get("version")
    if not version:
        fail("version.json missing 'version' field")
        return None

    if not re.match(r"^\d+\.\d+\.\d+$", version):
        fail(f"version.json version '{version}' doesn't match X.X.X format")
    else:
        ok(f"version.json version format valid: {version}")

    updated = data.get("updated")
    if not updated:
        fail("version.json missing 'updated' field")
    elif not re.match(r"^\d{4}-\d{2}-\d{2}$", updated):
        fail(f"version.json updated '{updated}' doesn't match YYYY-MM-DD format")
    else:
        ok(f"version.json updated date valid: {updated}")

    return version


# ---------------------------------------------------------------------------
# 2. APP_VERSION in index.html matches version.json
# ---------------------------------------------------------------------------
def check_app_version_sync(expected_version):
    print("\n--- APP_VERSION sync ---")

    if not os.path.exists(INDEX_HTML):
        fail("index.html not found")
        return

    with open(INDEX_HTML, "r") as f:
        content = f.read()

    match = re.search(r"const APP_VERSION\s*=\s*'([^']+)'", content)
    if not match:
        fail("Could not find APP_VERSION in index.html")
        return

    app_version = match.group(1)

    if expected_version and app_version != expected_version:
        fail(f"APP_VERSION '{app_version}' != version.json '{expected_version}'")
    elif expected_version:
        ok(f"APP_VERSION matches version.json: {app_version}")
    else:
        warn(f"APP_VERSION found ({app_version}) but couldn't verify against version.json")


# ---------------------------------------------------------------------------
# 3. HTML structure validation
# ---------------------------------------------------------------------------
def check_html_structure():
    print("\n--- HTML structure ---")

    if not os.path.exists(INDEX_HTML):
        fail("index.html not found")
        return

    with open(INDEX_HTML, "r") as f:
        content = f.read()

    # Check DOCTYPE
    if content.strip().startswith("<!DOCTYPE html>"):
        ok("DOCTYPE present")
    else:
        fail("Missing <!DOCTYPE html>")

    # Check essential tags exist
    for tag in ["<html", "</html>", "<head>", "</head>", "<body", "</body>"]:
        if tag in content:
            ok(f"Found {tag}")
        else:
            fail(f"Missing {tag}")

    # Check script and style tags are balanced
    style_opens = content.count("<style")
    style_closes = content.count("</style>")
    if style_opens == style_closes:
        ok(f"Style tags balanced ({style_opens} open, {style_closes} close)")
    else:
        fail(f"Style tags unbalanced ({style_opens} open, {style_closes} close)")

    script_opens = len(re.findall(r"<script[\s>]", content))
    script_closes = content.count("</script>")
    if script_opens == script_closes:
        ok(f"Script tags balanced ({script_opens} open, {script_closes} close)")
    else:
        fail(f"Script tags unbalanced ({script_opens} open, {script_closes} close)")


# ---------------------------------------------------------------------------
# 4. JS section headers check
# ---------------------------------------------------------------------------
def check_section_headers():
    print("\n--- JS section headers ---")

    if not os.path.exists(INDEX_HTML):
        fail("index.html not found")
        return

    with open(INDEX_HTML, "r") as f:
        lines = f.readlines()

    header_pattern = re.compile(r"//\s*={10,}.*={10,}")
    headers_found = []

    for i, line in enumerate(lines, 1):
        if header_pattern.search(line):
            header_text = line.strip()
            headers_found.append((i, header_text))

    if headers_found:
        ok(f"Found {len(headers_found)} JS section headers")
    else:
        warn("No section headers found — expected // ==================== patterns")


# ---------------------------------------------------------------------------
# 5. localStorage key consistency
# ---------------------------------------------------------------------------
def check_localstorage_keys():
    print("\n--- localStorage keys ---")

    if not os.path.exists(INDEX_HTML):
        fail("index.html not found")
        return

    content = read_all_js_content()

    # Find all localStorage key references across all files
    get_keys = set(re.findall(r"localStorage\.getItem\(['\"]([^'\"]+)['\"]\)", content))
    set_keys = set(re.findall(r"localStorage\.setItem\(['\"]([^'\"]+)['\"]\)", content))
    remove_keys = set(re.findall(r"localStorage\.removeItem\(['\"]([^'\"]+)['\"]\)", content))

    # Also find keys used via constants
    const_keys = set(re.findall(r"STORAGE_KEY_\w+", content))

    all_read = get_keys
    all_write = set_keys

    # Keys that are read but never written
    read_only = all_read - all_write - remove_keys
    if read_only:
        for key in read_only:
            warn(f"localStorage key '{key}' is read but never written")
    else:
        ok("All read localStorage keys have matching writes")

    # Keys that are written but never read
    write_only = all_write - all_read
    if write_only:
        for key in write_only:
            warn(f"localStorage key '{key}' is written but never read")
    else:
        ok("All written localStorage keys have matching reads")

    ok(f"Found {len(all_read | all_write)} unique localStorage keys")


# ---------------------------------------------------------------------------
# 6. Basic JS syntax checks
# ---------------------------------------------------------------------------
def check_js_syntax():
    print("\n--- JS basic syntax ---")

    if not os.path.exists(INDEX_HTML):
        fail("index.html not found")
        return

    with open(INDEX_HTML, "r") as f:
        content = f.read()

    # Extract JS content from inline <script> blocks
    js_blocks = re.findall(r"<script[^>]*>(.*?)</script>", content, re.DOTALL)

    if not js_blocks:
        fail("No script blocks found")
        return

    # Combine all inline JS + all extracted .js files
    all_js = "\n".join(js_blocks)
    for js_file in get_all_js_files():
        with open(js_file, "r") as f:
            all_js += "\n" + f.read()

    # Check bracket balance (rough check — doesn't handle strings/comments perfectly)
    # This is intentionally simple — a full parser would be overkill
    opens = all_js.count("{")
    closes = all_js.count("}")
    diff = opens - closes
    if abs(diff) <= 2:  # Small tolerance for template literals
        ok(f"Curly braces roughly balanced ({{={opens}, }}={closes}, diff={diff})")
    else:
        fail(f"Curly braces imbalanced ({{={opens}, }}={closes}, diff={diff})")

    paren_opens = all_js.count("(")
    paren_closes = all_js.count(")")
    paren_diff = paren_opens - paren_closes
    if abs(paren_diff) <= 2:
        ok(f"Parentheses roughly balanced ((={paren_opens}, )={paren_closes}, diff={paren_diff})")
    else:
        fail(f"Parentheses imbalanced ((={paren_opens}, )={paren_closes}, diff={paren_diff})")

    bracket_opens = all_js.count("[")
    bracket_closes = all_js.count("]")
    bracket_diff = bracket_opens - bracket_closes
    if abs(bracket_diff) <= 2:
        ok(f"Square brackets roughly balanced ([={bracket_opens}, ]={bracket_closes}, diff={bracket_diff})")
    else:
        fail(f"Square brackets imbalanced ([={bracket_opens}, ]={bracket_closes}, diff={bracket_diff})")

    # Per-file balance check for extracted modules
    js_files = get_all_js_files()
    bad_files = []
    for js_file in js_files:
        with open(js_file, "r") as f:
            fc = f.read()
        if abs(fc.count("{") - fc.count("}")) > 1 or \
           abs(fc.count("(") - fc.count(")")) > 1 or \
           abs(fc.count("[") - fc.count("]")) > 1:
            bad_files.append(os.path.relpath(js_file, PROJECT_ROOT))
    if bad_files:
        for bf in bad_files:
            fail(f"Bracket imbalance in {bf}")
    elif js_files:
        ok(f"All {len(js_files)} extracted JS modules have balanced brackets")


# ---------------------------------------------------------------------------
# 7. Dangerous patterns check
# ---------------------------------------------------------------------------
def check_dangerous_patterns():
    print("\n--- Dangerous patterns ---")

    if not os.path.exists(INDEX_HTML):
        fail("index.html not found")
        return

    dangerous = {
        r"\beval\s*\(": "eval() usage",
        r"\bnew\s+Function\s*\(": "Function() constructor",
        r"\bdocument\.write\s*\(": "document.write()",
    }

    # Collect all files to scan: index.html + extracted .js modules
    files_to_scan = [INDEX_HTML] + get_all_js_files()

    found_any = False
    for filepath in files_to_scan:
        if not os.path.exists(filepath):
            continue
        rel_path = os.path.relpath(filepath, PROJECT_ROOT)
        with open(filepath, "r") as f:
            lines = f.readlines()
        for pattern, desc in dangerous.items():
            for i, line in enumerate(lines, 1):
                # Skip comments
                stripped = line.strip()
                if stripped.startswith("//") or stripped.startswith("*"):
                    continue
                if re.search(pattern, line):
                    warn(f"{rel_path}:{i}: {desc} — {stripped[:80]}")
                    found_any = True

    if not found_any:
        ok("No eval(), Function(), or document.write() found")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------
def main():
    print("=" * 60)
    print("LED Wall Calculator — Smoke Test")
    print("=" * 60)

    version = check_version_json()
    check_app_version_sync(version)
    check_html_structure()
    check_section_headers()
    check_localstorage_keys()
    check_js_syntax()
    check_dangerous_patterns()

    print("\n" + "=" * 60)
    print(f"Results: {passed} passed, {failed} failed, {warnings} warnings")
    print("=" * 60)

    if failed > 0:
        sys.exit(1)
    else:
        sys.exit(0)


if __name__ == "__main__":
    main()

File: tests/smoke-test.js
/**
 * LED Wall Calculator — Smoke Test (Node.js)
 * Runs basic validation checks after code changes.
 * Usage: node tests/smoke-test.js
 *
 * This is a 1:1 port of tests/smoke-test.py for machines without Python3.
 */

const fs = require("fs");
const path = require("path");
const { execFileSync } = require("child_process");
const os = require("os");

const PROJECT_ROOT = path.resolve(__dirname, "..");
const INDEX_HTML = path.join(PROJECT_ROOT, "index.html");
const VERSION_JSON = path.join(PROJECT_ROOT, "version.json");
const SW_JS = path.join(PROJECT_ROOT, "sw.js");

const JS_MODULE_DIRS = [
  "core", "specs", "layouts", "structure", "interact",
  "nav", "export", "config", "screens"
];

let passed = 0;
let failed = 0;
let warnings = 0;

function ok(msg) { passed++; console.log("  PASS  " + msg); }
function fail(msg) { failed++; console.log("  FAIL  " + msg); }
function warn(msg) { warnings++; console.log("  WARN  " + msg); }

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function getAllJsFiles() {
  const files = [];
  // Include sw.js from project root
  if (fs.existsSync(SW_JS)) {
    files.push(SW_JS);
  }
  for (const d of JS_MODULE_DIRS) {
    const dirPath = path.join(PROJECT_ROOT, d);
    if (fs.existsSync(dirPath) && fs.statSync(dirPath).isDirectory()) {
      const entries = fs.readdirSync(dirPath).filter(f => f.endsWith(".js")).sort();
      for (const fname of entries) {
        files.push(path.join(dirPath, fname));
      }
    }
  }
  return files;
}

function readAllJsContent() {
  let content = "";
  if (fs.existsSync(INDEX_HTML)) {
    content = fs.readFileSync(INDEX_HTML, "utf8");
  }
  for (const jsFile of getAllJsFiles()) {
    content += "\n" + fs.readFileSync(jsFile, "utf8");
  }
  return content;
}

// ---------------------------------------------------------------------------
// 1. version.json validation
// ---------------------------------------------------------------------------
function checkVersionJson() {
  console.log("\n--- version.json ---");

  if (!fs.existsSync(VERSION_JSON)) {
    fail("version.json not found");
    return null;
  }

  let data;
  try {
    data = JSON.parse(fs.readFileSync(VERSION_JSON, "utf8"));
  } catch (e) {
    fail("version.json is not valid JSON: " + e.message);
    return null;
  }
  ok("version.json is valid JSON");

  const version = data.version;
  if (!version) {
    fail("version.json missing 'version' field");
    return null;
  }
  if (!/^\d+\.\d+\.\d+$/.test(version)) {
    fail("version.json version '" + version + "' doesn't match X.X.X format");
  } else {
    ok("version.json version format valid: " + version);
  }

  const updated = data.updated;
  if (!updated) {
    fail("version.json missing 'updated' field");
  } else if (!/^\d{4}-\d{2}-\d{2}$/.test(updated)) {
    fail("version.json updated '" + updated + "' doesn't match YYYY-MM-DD format");
  } else {
    ok("version.json updated date valid: " + updated);
  }

  return version;
}

// ---------------------------------------------------------------------------
// 2. APP_VERSION in index.html matches version.json
// ---------------------------------------------------------------------------
function checkAppVersionSync(expectedVersion) {
  console.log("\n--- APP_VERSION sync ---");

  if (!fs.existsSync(INDEX_HTML)) {
    fail("index.html not found");
    return;
  }

  const content = fs.readFileSync(INDEX_HTML, "utf8");
  const match = content.match(/const APP_VERSION\s*=\s*'([^']+)'/);
  if (!match) {
    fail("Could not find APP_VERSION in index.html");
    return;
  }

  const appVersion = match[1];
  if (expectedVersion && appVersion !== expectedVersion) {
    fail("APP_VERSION '" + appVersion + "' != version.json '" + expectedVersion + "'");
  } else if (expectedVersion) {
    ok("APP_VERSION matches version.json: " + appVersion);
  } else {
    warn("APP_VERSION found (" + appVersion + ") but couldn't verify against version.json");
  }
}

// ---------------------------------------------------------------------------
// 2b. SW_VERSION in sw.js matches version.json
// ---------------------------------------------------------------------------
function checkSwVersionSync(expectedVersion) {
  console.log("\n--- SW_VERSION sync ---");

  if (!fs.existsSync(SW_JS)) {
    fail("sw.js not found");
    return;
  }

  const content = fs.readFileSync(SW_JS, "utf8");
  const match = content.match(/const SW_VERSION\s*=\s*'([^']+)'/);
  if (!match) {
    fail("Could not find SW_VERSION in sw.js");
    return;
  }

  const swVersion = match[1];
  if (expectedVersion && swVersion !== expectedVersion) {
    fail("SW_VERSION '" + swVersion + "' != version.json '" + expectedVersion + "'");
  } else if (expectedVersion) {
    ok("SW_VERSION matches version.json: " + swVersion);
  } else {
    warn("SW_VERSION found (" + swVersion + ") but couldn't verify against version.json");
  }
}

// ---------------------------------------------------------------------------
// 3. HTML structure validation
// ---------------------------------------------------------------------------
function checkHtmlStructure() {
  console.log("\n--- HTML structure ---");

  if (!fs.existsSync(INDEX_HTML)) {
    fail("index.html not found");
    return;
  }

  const content = fs.readFileSync(INDEX_HTML, "utf8");

  // DOCTYPE
  if (content.trimStart().startsWith("<!DOCTYPE html>")) {
    ok("DOCTYPE present");
  } else {
    fail("Missing <!DOCTYPE html>");
  }

  // Essential tags
  for (const tag of ["<html", "</html>", "<head>", "</head>", "<body", "</body>"]) {
    if (content.includes(tag)) {
      ok("Found " + tag);
    } else {
      fail("Missing " + tag);
    }
  }

  // Balanced style tags
  const styleOpens = (content.match(/<style/g) || []).length;
  const styleCloses = (content.match(/<\/style>/g) || []).length;
  if (styleOpens === styleCloses) {
    ok("Style tags balanced (" + styleOpens + " open, " + styleCloses + " close)");
  } else {
    fail("Style tags unbalanced (" + styleOpens + " open, " + styleCloses + " close)");
  }

  // Balanced script tags
  const scriptOpens = (content.match(/<script[\s>]/g) || []).length;
  const scriptCloses = (content.match(/<\/script>/g) || []).length;
  if (scriptOpens === scriptCloses) {
    ok("Script tags balanced (" + scriptOpens + " open, " + scriptCloses + " close)");
  } else {
    fail("Script tags unbalanced (" + scriptOpens + " open, " + scriptCloses + " close)");
  }
}

// ---------------------------------------------------------------------------
// 4. JS section headers check
// ---------------------------------------------------------------------------
function checkSectionHeaders() {
  console.log("\n--- JS section headers ---");

  if (!fs.existsSync(INDEX_HTML)) {
    fail("index.html not found");
    return;
  }

  const lines = fs.readFileSync(INDEX_HTML, "utf8").split("\n");
  const headerPattern = /\/\/\s*={10,}.*={10,}/;
  let headersFound = 0;

  for (const line of lines) {
    if (headerPattern.test(line)) headersFound++;
  }

  if (headersFound > 0) {
    ok("Found " + headersFound + " JS section headers");
  } else {
    warn("No section headers found — expected // ==================== patterns");
  }
}

// ---------------------------------------------------------------------------
// 5. localStorage key consistency
// ---------------------------------------------------------------------------
function checkLocalStorageKeys() {
  console.log("\n--- localStorage keys ---");

  const content = readAllJsContent();

  const getKeys = new Set((content.match(/localStorage\.getItem\(['"]([^'"]+)['"]\)/g) || [])
    .map(m => m.match(/['"]([^'"]+)['"]/)[1]));
  const setKeys = new Set((content.match(/localStorage\.setItem\(['"]([^'"]+)['"]\)/g) || [])
    .map(m => m.match(/['"]([^'"]+)['"]/)[1]));
  const removeKeys = new Set((content.match(/localStorage\.removeItem\(['"]([^'"]+)['"]\)/g) || [])
    .map(m => m.match(/['"]([^'"]+)['"]/)[1]));

  // Keys read but never written
  const readOnly = [...getKeys].filter(k => !setKeys.has(k) && !removeKeys.has(k));
  if (readOnly.length > 0) {
    for (const key of readOnly) {
      warn("localStorage key '" + key + "' is read but never written");
    }
  } else {
    ok("All read localStorage keys have matching writes");
  }

  // Keys written but never read
  const writeOnly = [...setKeys].filter(k => !getKeys.has(k));
  if (writeOnly.length > 0) {
    for (const key of writeOnly) {
      warn("localStorage key '" + key + "' is written but never read");
    }
  } else {
    ok("All written localStorage keys have matching reads");
  }

  const allKeys = new Set([...getKeys, ...setKeys]);
  ok("Found " + allKeys.size + " unique localStorage keys");
}

// ---------------------------------------------------------------------------
// 6. Basic JS syntax checks (bracket balance)
// ---------------------------------------------------------------------------
function checkJsSyntax() {
  console.log("\n--- JS basic syntax ---");

  if (!fs.existsSync(INDEX_HTML)) {
    fail("index.html not found");
    return;
  }

  const content = fs.readFileSync(INDEX_HTML, "utf8");

  // Extract inline JS blocks
  const jsBlocks = [];
  const blockRegex = /<script[^>]*>([\s\S]*?)<\/script>/g;
  let m;
  while ((m = blockRegex.exec(content)) !== null) {
    // Skip external script tags (those with src=)
    const tagOpen = m[0].substring(0, m[0].indexOf(">"));
    if (tagOpen.includes("src=")) continue;
    jsBlocks.push(m[1]);
  }

  // Combine all inline JS + all extracted .js files
  let allJs = jsBlocks.join("\n");
  for (const jsFile of getAllJsFiles()) {
    allJs += "\n" + fs.readFileSync(jsFile, "utf8");
  }

  // Global bracket balance
  function countChar(str, ch) {
    let count = 0;
    for (let i = 0; i < str.length; i++) {
      if (str[i] === ch) count++;
    }
    return count;
  }

  const curlyOpen = countChar(allJs, "{");
  const curlyClose = countChar(allJs, "}");
  const curlyDiff = curlyOpen - curlyClose;
  if (Math.abs(curlyDiff) <= 2) {
    ok("Curly braces roughly balanced ({=" + curlyOpen + ", }=" + curlyClose + ", diff=" + curlyDiff + ")");
  } else {
    fail("Curly braces imbalanced ({=" + curlyOpen + ", }=" + curlyClose + ", diff=" + curlyDiff + ")");
  }

  const parenOpen = countChar(allJs, "(");
  const parenClose = countChar(allJs, ")");
  const parenDiff = parenOpen - parenClose;
  if (Math.abs(parenDiff) <= 2) {
    ok("Parentheses roughly balanced ((=" + parenOpen + ", )=" + parenClose + ", diff=" + parenDiff + ")");
  } else {
    fail("Parentheses imbalanced ((=" + parenOpen + ", )=" + parenClose + ", diff=" + parenDiff + ")");
  }

  const bracketOpen = countChar(allJs, "[");
  const bracketClose = countChar(allJs, "]");
  const bracketDiff = bracketOpen - bracketClose;
  if (Math.abs(bracketDiff) <= 2) {
    ok("Square brackets roughly balanced ([=" + bracketOpen + ", ]=" + bracketClose + ", diff=" + bracketDiff + ")");
  } else {
    fail("Square brackets imbalanced ([=" + bracketOpen + ", ]=" + bracketClose + ", diff=" + bracketDiff + ")");
  }

  // Per-file balance check
  const jsFiles = getAllJsFiles();
  const badFiles = [];
  for (const jsFile of jsFiles) {
    const fc = fs.readFileSync(jsFile, "utf8");
    if (Math.abs(countChar(fc, "{") - countChar(fc, "}")) > 1 ||
        Math.abs(countChar(fc, "(") - countChar(fc, ")")) > 1 ||
        Math.abs(countChar(fc, "[") - countChar(fc, "]")) > 1) {
      badFiles.push(path.relative(PROJECT_ROOT, jsFile));
    }
  }
  if (badFiles.length > 0) {
    for (const bf of badFiles) {
      fail("Bracket imbalance in " + bf);
    }
  } else if (jsFiles.length > 0) {
    ok("All " + jsFiles.length + " extracted JS modules have balanced brackets");
  }
}

// ---------------------------------------------------------------------------
// 7. JS parse validation via Node.js new Function()
// ---------------------------------------------------------------------------
function checkJsParse() {
  console.log("\n--- JS parse validation (Node.js) ---");

  if (!fs.existsSync(INDEX_HTML)) {
    fail("index.html not found");
    return;
  }

  const content = fs.readFileSync(INDEX_HTML, "utf8");

  // Extract inline script blocks (skip src= external scripts)
  const blocks = [];
  const blockRegex = /<script([^>]*)>([\s\S]*?)<\/script>/g;
  let m;
  while ((m = blockRegex.exec(content)) !== null) {
    if (m[1].includes("src=")) continue;
    blocks.push(m[2]);
  }

  // Validate each inline block
  let inlineOk = true;
  for (let i = 0; i < blocks.length; i++) {
    try {
      new Function(blocks[i]);
    } catch (e) {
      fail("Inline script block " + i + ": " + e.message);
      inlineOk = false;
    }
  }
  if (inlineOk && blocks.length > 0) {
    ok("All " + blocks.length + " inline script blocks parse successfully");
  }

  // Validate each extracted .js module
  const jsFiles = getAllJsFiles();
  const bad = [];
  for (const jsFile of jsFiles) {
    const rel = path.relative(PROJECT_ROOT, jsFile);
    try {
      const src = fs.readFileSync(jsFile, "utf8");
      new Function(src);
    } catch (e) {
      fail(rel + ": " + e.message);
      bad.push(rel);
    }
  }
  if (bad.length === 0 && jsFiles.length > 0) {
    ok("All " + jsFiles.length + " extracted JS modules parse successfully");
  }
}

// ---------------------------------------------------------------------------
// 8. Dangerous patterns check
// ---------------------------------------------------------------------------
function checkDangerousPatterns() {
  console.log("\n--- Dangerous patterns ---");

  const dangerous = [
    { pattern: /\beval\s*\(/, desc: "eval() usage" },
    { pattern: /\bnew\s+Function\s*\(/, desc: "Function() constructor" },
    { pattern: /\bdocument\.write\s*\(/, desc: "document.write()" }
  ];

  const filesToScan = [INDEX_HTML, ...getAllJsFiles()];
  let foundAny = false;

  for (const filepath of filesToScan) {
    if (!fs.existsSync(filepath)) continue;
    const relPath = path.relative(PROJECT_ROOT, filepath);
    const lines = fs.readFileSync(filepath, "utf8").split("\n");

    for (let i = 0; i < lines.length; i++) {
      const stripped = lines[i].trim();
      // Skip comments
      if (stripped.startsWith("//") || stripped.startsWith("*")) continue;

      for (const { pattern, desc } of dangerous) {
        if (pattern.test(lines[i])) {
          warn(relPath + ":" + (i + 1) + ": " + desc + " — " + stripped.substring(0, 80));
          foundAny = true;
        }
      }
    }
  }

  if (!foundAny) {
    ok("No eval(), Function(), or document.write() found");
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------
console.log("=".repeat(60));
console.log("LED Wall Calculator — Smoke Test (Node.js)");
console.log("=".repeat(60));

const version = checkVersionJson();
checkAppVersionSync(version);
checkSwVersionSync(version);
checkHtmlStructure();
checkSectionHeaders();
checkLocalStorageKeys();
checkJsSyntax();
checkJsParse();
checkDangerousPatterns();

console.log("\n" + "=".repeat(60));
console.log("Results: " + passed + " passed, " + failed + " failed, " + warnings + " warnings");
console.log("=".repeat(60));

process.exit(failed > 0 ? 1 : 0);
